File: Lib/Telerik/Examples/ExportWithDpl/SpreadsheetStreaming/ExampleViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Data;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using Telerik.Documents.SpreadsheetStreaming;
using Telerik.WinControls.Export;
using Telerik.WinControls.UI;

namespace ExportWithDpl.SpreadsheetStreaming
{
    public class ExampleViewModel : INotifyPropertyChanged
    {
        private static readonly int ColumnsCount = 8;

        private int exportedCellsCount;
        private double exportTime;
        private int rowIndex = -1;
        private int rowsCount;
        private SpreadStreamExportFormat selectedDocumentFormat;
        private int totalCellsCount;
        private DateTime exportStarted;
        private bool canExport;
        private long memory;
        private BindingSource data;
        private List<DocumentRow> rows;

        public ExampleViewModel()
        {
            this.data = new BindingSource();
            this.rows = new List<DocumentRow>();
            this.data.DataSource = this.rows;
            this.RowsCount = 5000;
            this.CanExport = true;

            this.AddRemoveData();
        }

        public BindingSource Data 
        {
            get { return this.data; }
        }

        public SpreadStreamExportFormat SelectedDocumentFormat
        {
            get
            {
                return this.selectedDocumentFormat;
            }
            set
            {
                if (this.selectedDocumentFormat != value)
                {
                    this.selectedDocumentFormat = value;
                    this.OnPropertyChanged("SelectedDocumentFormat");
                }
            }
        }

        public IEnumerable<SpreadStreamExportFormat> ExportFormats
        {
            get { return Enum.GetValues(typeof(SpreadStreamExportFormat)).Cast<SpreadStreamExportFormat>(); }
        }

        public int RowsCount
        {
            get
            {
                return this.rowsCount;
            }
            set
            {
                if (this.rowsCount != value)
                {
                    this.rowsCount = value;
                    this.ExportedCellsCount = 0;
                    this.TotalCellsCount = ColumnsCount * value;
                    this.AddRemoveData();
                    this.OnPropertyChanged("RowsCount");
                }
            }
        }

        public int ExportedCellsCount
        {
            get
            {
                return this.exportedCellsCount;
            }
            set
            {
                if (this.exportedCellsCount != value)
                {
                    this.exportedCellsCount = value;
                    this.OnPropertyChanged("ExportedCellsCount");
                }
            }
        }

        public int TotalCellsCount
        {
            get
            {
                return this.totalCellsCount;
            }
            set
            {
                if (this.totalCellsCount != value)
                {
                    this.totalCellsCount = value;
                    this.OnPropertyChanged("TotalCellsCount");
                }
            }
        }

        public double ExportTime
        {
            get
            {
                return this.exportTime;
            }
            set
            {
                if (this.exportTime != value)
                {
                    this.exportTime = value;
                    this.OnPropertyChanged("ExportTime");
                }
            }
        }

        public long Memory
        {
            get { return memory; }
            set
            {
                if (this.memory != value)
                {
                    this.memory = value;
                    this.OnPropertyChanged("Memory");
                }
            }
        }

        public bool CanExport
        {
            get
            {
                return this.canExport;
            }
            set
            {
                if (this.canExport != value)
                {
                    this.canExport = value;
                    this.OnPropertyChanged("CanExport");
                }
            }
        }

        private void AddRemoveData()
        {
            this.Data.SuspendBinding();

            while (this.rows.Count < this.RowsCount)
            {
                this.rows.Add(DocumentHelper.GenerateDocumentRow(rowsCount, this.rows.Count));
            }

            while (this.rows.Count > this.RowsCount)
            {
                this.rows.RemoveAt(this.rows.Count - 1);
            }

            this.Data.ResumeBinding();
            this.Data.ResetBindings(false);
        }

        public void Export(RadGridView gridToExport)
        {
            SaveFileDialog dialog = new SaveFileDialog();

            dialog.Filter = String.Format("{0} files|*.{1}|All files (*.*)|*.*",
                this.SelectedDocumentFormat.ToString().ToLower(),
                this.SelectedDocumentFormat.ToString().ToLower());
            dialog.FileName = "GridViewSpreadStreamExport";
            if (dialog.ShowDialog() == DialogResult.OK)
            {
                this.CanExport = false;
                this.ExportedCellsCount = 0;
                this.ExportTime = 0;
                this.exportStarted = DateTime.Now;
                rowIndex = -1;

                GridViewSpreadStreamExport spreadStreamExport = new GridViewSpreadStreamExport(gridToExport);
                spreadStreamExport.ExportFormat = this.SelectedDocumentFormat;
                spreadStreamExport.RowCreated += spreadStreamExport_RowCreated;
                spreadStreamExport.RunExport(dialog.FileName, new SpreadStreamExportRenderer());
                this.CanExport = true;
            }
        }

        private void spreadStreamExport_RowCreated(object sender, SpreadStreamRowEventArgs e)
        {
            rowIndex += 1;

            if (rowIndex % (this.RowsCount / 100) == 0 || rowIndex == this.RowsCount - 1)
            {
                this.ExportTime = (DateTime.Now - this.exportStarted).TotalMilliseconds / 1000;
                this.ExportedCellsCount = ColumnsCount * rowIndex;
                this.Memory = GC.GetTotalMemory(false) / 1024 / 1024;
                Application.DoEvents();
            }
        }

        public event PropertyChangedEventHandler PropertyChanged;

        protected virtual void OnPropertyChanged(string propertyName)
        {
            PropertyChangedEventHandler handler = this.PropertyChanged;

            if (handler != null)
            {
                handler(this, new PropertyChangedEventArgs(propertyName));
            }
        }
    }
}

File: Lib/Telerik/Examples/ExportWithDpl/SpreadsheetStreaming/ExampleViewModel.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Data
Imports System.IO
Imports System.Linq
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports Telerik.Documents.SpreadsheetStreaming
Imports Telerik.WinControls.Export
Imports Telerik.WinControls.UI

Namespace ExportWithDpl.SpreadsheetStreaming
    Public Class ExampleViewModel
        Implements INotifyPropertyChanged
        Private Shared ReadOnly ColumnsCount As Integer = 8

        Private exportedCellsCount_Renamed As Integer
        Private exportTime_Renamed As Double
        Private rowIndex As Integer = -1
        Private rowsCount_Renamed As Integer
        Private selectedDocumentFormat_Renamed As SpreadStreamExportFormat
        Private totalCellsCount_Renamed As Integer
        Private exportStarted As DateTime
        Private canExport_Renamed As Boolean
        Private memory_Renamed As Long
        Private data_Renamed As BindingSource
        Private rows As List(Of DocumentRow)

        Public Sub New()
            Me.data_Renamed = New BindingSource()
            Me.rows = New List(Of DocumentRow)()
            Me.data_Renamed.DataSource = Me.rows
            Me.RowsCount = 5000
            Me.CanExport = True

            Me.AddRemoveData()
        End Sub

        Public ReadOnly Property Data() As BindingSource
            Get
                Return Me.data_Renamed
            End Get
        End Property

        Public Property SelectedDocumentFormat() As SpreadStreamExportFormat
            Get
                Return Me.selectedDocumentFormat_Renamed
            End Get
            Set(value As SpreadStreamExportFormat)
                If Not Me.selectedDocumentFormat_Renamed <> Value Then
                    Me.selectedDocumentFormat_Renamed = Value
                    Me.OnPropertyChanged("SelectedDocumentFormat")
                End If
            End Set
        End Property

        Public ReadOnly Property ExportFormats() As IEnumerable(Of SpreadStreamExportFormat)
            Get
                Return System.Enum.GetValues(GetType(SpreadStreamExportFormat)).Cast(Of SpreadStreamExportFormat)()
            End Get
        End Property

        Public Property RowsCount() As Integer
            Get
                Return Me.rowsCount_Renamed
            End Get
            Set(value As Integer)
                If Me.rowsCount_Renamed <> Value Then
                    Me.rowsCount_Renamed = Value
                    Me.ExportedCellsCount = 0
                    Me.TotalCellsCount = ColumnsCount * Value
                    Me.AddRemoveData()
                    Me.OnPropertyChanged("RowsCount")
                End If
            End Set
        End Property

        Public Property ExportedCellsCount() As Integer
            Get
                Return Me.exportedCellsCount_Renamed
            End Get
            Set(value As Integer)
                If Me.exportedCellsCount_Renamed <> Value Then
                    Me.exportedCellsCount_Renamed = Value
                    Me.OnPropertyChanged("ExportedCellsCount")
                End If
            End Set
        End Property

        Public Property TotalCellsCount() As Integer
            Get
                Return Me.totalCellsCount_Renamed
            End Get
            Set(value As Integer)
                If Me.totalCellsCount_Renamed <> Value Then
                    Me.totalCellsCount_Renamed = Value
                    Me.OnPropertyChanged("TotalCellsCount")
                End If
            End Set
        End Property

        Public Property ExportTime() As Double
            Get
                Return Me.exportTime_Renamed
            End Get
            Set(value As Double)
                If Me.exportTime_Renamed <> Value Then
                    Me.exportTime_Renamed = Value
                    Me.OnPropertyChanged("ExportTime")
                End If
            End Set
        End Property

        Public Property Memory() As Long
            Get
                Return memory_Renamed
            End Get
            Set(value As Long)
                If Me.memory_Renamed <> Value Then
                    Me.memory_Renamed = Value
                    Me.OnPropertyChanged("Memory")
                End If
            End Set
        End Property

        Public Property CanExport() As Boolean
            Get
                Return Me.canExport_Renamed
            End Get
            Set(value As Boolean)
                If Me.canExport_Renamed <> Value Then
                    Me.canExport_Renamed = Value
                    Me.OnPropertyChanged("CanExport")
                End If
            End Set
        End Property

        Private Sub AddRemoveData()
            Me.Data.SuspendBinding()

            Do While Me.rows.Count < Me.RowsCount
                Me.rows.Add(DocumentHelper.GenerateDocumentRow(rowsCount_Renamed, Me.rows.Count))
            Loop

            Do While Me.rows.Count > Me.RowsCount
                Me.rows.RemoveAt(Me.rows.Count - 1)
            Loop

            Me.Data.ResumeBinding()
            Me.Data.ResetBindings(False)
        End Sub

        Public Sub Export(ByVal gridToExport As RadGridView)
            Dim dialog As SaveFileDialog = New SaveFileDialog()

            dialog.Filter = String.Format("{0} files|*.{1}|All files (*.*)|*.*", Me.SelectedDocumentFormat.ToString().ToLower(), Me.SelectedDocumentFormat.ToString().ToLower())
            dialog.FileName = "GridViewSpreadStreamExport"
            If dialog.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
                Me.CanExport = False
                Me.ExportedCellsCount = 0
                Me.ExportTime = 0
                Me.exportStarted = DateTime.Now
                rowIndex = -1

                Dim spreadStreamExport As GridViewSpreadStreamExport = New GridViewSpreadStreamExport(gridToExport)
                AddHandler spreadStreamExport.RowCreated, AddressOf spreadStreamExport_RowCreated
                spreadStreamExport.RunExport(dialog.FileName, New SpreadStreamExportRenderer())
                Me.CanExport = True
            End If
        End Sub

        Private Sub spreadStreamExport_RowCreated(ByVal sender As Object, ByVal e As SpreadStreamRowEventArgs)
            rowIndex += 1

            If rowIndex Mod (Me.RowsCount / 100) = 0 OrElse rowIndex = Me.RowsCount - 1 Then
                Me.ExportTime = (DateTime.Now - Me.exportStarted).TotalMilliseconds / 1000
                Me.ExportedCellsCount = ColumnsCount * rowIndex
                Me.Memory = GC.GetTotalMemory(False) / 1024L / 1024L
                Application.DoEvents()
            End If
        End Sub

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Protected Overridable Sub OnPropertyChanged(ByVal propertyName As String)
            Dim handler As PropertyChangedEventHandler = Me.PropertyChangedEvent

            If Not handler Is Nothing Then
                handler(Me, New PropertyChangedEventArgs(propertyName))
            End If
        End Sub
    End Class
End Namespace
